### PR TITLE
[ASCornerLayoutSpec] New layout spec class for declarative corner element layout.

### PR DESCRIPTION
**Pull request by huang-kun**
This PR includes implementation code, snapshot test and examples (both objc and swift).

### CODE REVIEW

**Comment by huang-kun** [comment]
This PR comes from [#623](https://github.com/TextureGroup/Texture/pull/632). I will start one commit here instead of a whole mass. Make it easy for code review and CI build.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/Layout/ASCornerLayoutSpec.mm [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  
+  // Update corner's position
+  cornerLayout.position = relativePosition;
+  corner.style.layoutPosition = relativePosition;
+  
+  // Calculate size
+  CGRect frame = childLayout.frame;
+  if (_includeCornerForSizeCalculation) {
+    frame = CGRectUnion(childLayout.frame, cornerLayout.frame);
+    frame.size = ASSizeRangeClamp(constrainedSize, frame.size);
```
Comment (nguyenhuy) [inline Source/Layout/ASCornerLayoutSpec.mm]: Yes, always clamp against the constrained size range 👍

File: Tests/ASCornerLayoutSpecSnapshotTests.mm [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+  [self testCornerSpecWithLocation:ASCornerLayoutLocationBottomLeft offsetOption:outer childSizeOnly:NO];
+  [self testCornerSpecWithLocation:ASCornerLayoutLocationBottomLeft offsetOption:outer childSizeOnly:YES];
+
+  [self testCornerSpecWithLocation:ASCornerLayoutLocationBottomRight offsetOption:outer childSizeOnly:NO];
+  [self testCornerSpecWithLocation:ASCornerLayoutLocationBottomRight offsetOption:outer childSizeOnly:YES];
+}
+
+- (void)testCornerSpecWithLocation:(ASCornerLayoutLocation)location
+                      offsetOption:(ASCornerLayoutSpecSnapshotTestsOffsetOption)offsetOption
+                     childSizeOnly:(BOOL)childSizeOnly {
```
Comment (nguyenhuy) [inline Tests/ASCornerLayoutSpecSnapshotTests.mm]: Let's use the same name that the corner spec uses (`includeCornerForSizeCalculation` or `wrapsCorner` depends on its final name).

File: examples/LayoutSpecExamples-Swift/Sample/LayoutExampleNode.swift [hunk 3]
```diff
@@ -143,3 +143,144 @@ class FlexibleSeparatorSurroundingContent : LayoutExampleNode {
...
+    circularize(badgeNode, in: ImageSize.badge)
+    
+    makeSingleLine(for: usernameNode, with: "Hello world", fontSize: 17, textColor: .black)
+    makeSingleLine(for: subtitleNode, with: "This is a long long subtitle, with a long long appended string.", fontSize: 14, textColor: .lightGray)
+  }
+  
+  private func circularize(_ node: ASDisplayNode, in size: CGSize) {
+    assert(size.width == size.height)
+    node.style.preferredSize = size
+    node.cornerRadius = size.width / 2
```
Comment (nguyenhuy) [inline examples/LayoutSpecExamples-Swift/Sample/LayoutExampleNode.swift]: Not a blocker and don't spend too much time on this, but please consider other rounding techniques mentioned [here](http://texturegroup.org/docs/corner-rounding.html), if possible. I'm not concerned about the performance in this sample application, but rather using best practices in example code.

File: examples/LayoutSpecExamples/Sample/LayoutExampleNodes.m [hunk 4]
```diff
@@ -260,6 +260,169 @@ - (ASLayoutSpec *)layoutSpecThatFits:(ASSizeRange)constrainedSize
...
+        _dotNode = [ASImageNode new];
+        _dotNode.image = cornerImage;
+        
+        _photoNode1 = [ASImageNode new];
+        _photoNode1.image = avatarImage;
+        
+        _badgeNode = [ASTextNode new];
+        _badgeNode.attributedText = numberText;
+        _badgeNode.backgroundColor = UIColor.redColor;
+        _badgeNode.cornerRadius = 12;
```
Comment (nguyenhuy) [inline examples/LayoutSpecExamples/Sample/LayoutExampleNodes.m]: Same here and below, let's avoid `cornerRadius` whenever makes sense.

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Source/Layout/ASCornerLayoutSpec.mm [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+
+#import <AsyncDisplayKit/ASCornerLayoutSpec.h>
+#import <AsyncDisplayKit/ASLayout.h>
+#import <AsyncDisplayKit/ASLayoutSpec+Subclasses.h>
+#import <AsyncDisplayKit/ASDisplayNode.h>
+
+CGPoint as_calculatedCornerOriginIn(CGRect baseFrame, CGSize cornerSize, ASCornerLayoutLocation cornerLocation, CGPoint offset) {
```
Comment (nguyenhuy) [inline Source/Layout/ASCornerLayoutSpec.mm]: Nit: open curly bracket in a new line.

File: Tests/ASCornerLayoutSpecSnapshotTests.mm [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+@property (nonatomic, assign) CGSize contextSize;
+
+@property (nonatomic, assign) ASSizeRange contextSizeRange;
+
+@end
+
+
+@implementation ASCornerLayoutSpecSnapshotTests
+
+- (void)setUp {
```
Comment (nguyenhuy) [inline Tests/ASCornerLayoutSpecSnapshotTests.mm]: Nit: method curly brackets in new lines throughout this file.

File: Source/Layout/ASCornerLayoutSpec.mm [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+- (id<ASLayoutElement>)corner
+{
+  return [super childAtIndex:kCornerChildIndex];
+}
+
+#pragma mark - Calculation
+
+- (ASLayout *)calculateLayoutThatFits:(ASSizeRange)constrainedSize
```
Comment (nguyenhuy) [inline Source/Layout/ASCornerLayoutSpec.mm]: This whole layout algorithm is a beauty!

File: Source/Layout/ASCornerLayoutSpec.mm (hunk 5 above)
Comment (huang-kun) [reply Source/Layout/ASCornerLayoutSpec.mm]: Thank you for refining so much details on the new layout spec. I kept the curly bracket at the end of function as the code guide line declared. 🤔🤔🤔 It says:

* Function brackets on the same line

```
static void someFunction() {
  // Implementation
}
```

File: Source/Layout/ASCornerLayoutSpec.mm (hunk 7 above)
Comment (huang-kun) [reply Source/Layout/ASCornerLayoutSpec.mm]: Great to hear that 😁

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by pinterest-ios** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This is a big PR, please consider splitting it up to ease code review.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by nguyenhuy** [comment]
Here we go. Thanks again!

**Comment by huang-kun** [comment]
Would you like me to introduce the ASCornerLayoutSpec into your official doc at [here](https://github.com/TextureGroup/Texture/blob/master/docs/_docs/layout2-layoutspec-types.md) after new version release? @nguyenhuy @maicki

**Comment by nguyenhuy** [comment]
Yes, that would be great!

-- 
Best regards,
Huy Nguyen

**Pull request merged**